attest/core/src/ias/verifier.rs

```diff
@@ -0,0 +1,1138 @@
+            }));
+        self
+    }
+
+    /// Verify the report body (enclave) version is at least the given value.
+    pub fn version(&mut self, version: SecurityVersion) -> &mut Self {
+        self.report_body_verifiers
+            .push(VerifyReportBodyType::Version(VersionVerifier { version }));
+        self
+    }
+
+    /// Verify the given MrEnclave-based status verifier succeeds
+    pub fn mr_enclave(&mut self, verifier: MrEnclaveVerifier) -> &mut Self {
+        self.status_verifiers
+            .push(VerifyIasReportDataType::Enclave(verifier));
+        self
+    }
+
+    /// Verify the given MrSigner-based status verifier succeeds
+    pub fn mr_signer(&mut self, verifier: MrSignerVerifier) -> &mut Self {
+        self.status_verifiers
+            .push(VerifyIasReportDataType::Signer(verifier));
+        self
+    }
+
+    /// Compile the report verifier which a report will be given to
+    pub fn generate(&mut self) -> IasReportVerifier {
+        // Build a list of quote verifiers
+        let mut quote_verifiers = self.quote_verifiers.clone();
+        quote_verifiers.push(VerifyQuoteType::ReportBody(ReportBodyVerifier {
+            body_verifiers: self.report_body_verifiers.clone(),
+        }));
+
+        // Build the list of IAS report data verifiers (including a quote
+        // verifier)
+        let mut verifiers = self.status_verifiers.clone();
+        verifiers.push(VerifyIasReportDataType::Quote(QuoteVerifier {
+            quote_verifiers,
+        }));
+        verifiers.extend_from_slice(&self.ias_verifiers);
+
+        // Construct the top-level verifier.
+        IasReportVerifier {
+            trust_anchors: self.trust_anchors.clone(),
+            verifiers,
+        }
+    }
+}
+
+/// A structure which can verify a top-level report.
+pub struct IasReportVerifier {
+    /// A vector of trust anchor certificates to verify the report signature and
+    /// chain against.
+    trust_anchors: Vec<Certificate>,
+    /// A list of report data verifiers to be applied to the contents of the
+    /// JSON report.
+    verifiers: Vec<VerifyIasReportDataType>,
+}
+
+/// An enumeration of errors which an [`IasReportVerifier`] can produce.
+#[derive(Clone, Debug, Display, PartialEq, PartialOrd)]
+pub enum VerifierError {
+    /// The IAS report does not contain a certificate chain.
+    NoChain,
+    /// The signature is invalid, or was produced by a public key we do not
+    /// trust.
+    BadSignature,
+    /// There was an error parsing the JSON contents
+    Parse(VerifyError),
+    /// The report was properly constructed did not meet security requirements.
+    Verification(VerificationReportData),
+}
+
+impl From<VerifyError> for VerifierError {
+    fn from(src: VerifyError) -> VerifierError {
+        VerifierError::Parse(src)
+    }
+}
+
+const MAX_CHAIN_DEPTH: usize = 5;
+
+impl IasReportVerifier {
+    /// Verify the given IAS report using this verifier object.
+    pub fn verify(
```

Comment:
How difficult would it be to provide some unit test coverage for this (or is this functionality tested elsewhere)?